elyra/pipeline/validate.py

```diff
@@ -0,0 +1,633 @@
+        return self._has_fatal
+
+    def add_message(self,
+                    message: str,
+                    message_type: Optional[str] = "",
+                    data: Optional[Dict] = "",
+                    severity: ValidationSeverity = ValidationSeverity.Warning):
+        """
+        Helper function to add a diagnostic message to the response to be sent back
+        :param message: A simple message describing the issue
+        :param message_type: The type of message to send back e.g. invalidNodeType, invalidPipeline
+        :param data: a Dict with granular details regarding the error e.g. the nodeID, pipelineID, linkID etc.
+        :param severity: the severity level of the issue
+        :return:
+        """
+        valid_severity_levels = [ValidationSeverity.Error, ValidationSeverity.Warning,
+                                 ValidationSeverity.Information, ValidationSeverity.Hint]
+
+        if severity in valid_severity_levels:
+            diagnostic = {"severity": severity.value,
+                          "source": "Elyra Pipeline Validation Service",
+                          "type": message_type,
+                          "message": message,
+                          "data": data
+                          }
+            self._response['issues'].append(diagnostic)
+
+        if severity is ValidationSeverity.Error:
+            self._has_fatal = True
+
+    def to_json(self):
+        return self._response
+
+
+class PipelineValidationManager(SingletonConfigurable):
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.root_dir = get_expanded_path(kwargs.get('root_dir'))
+
+    async def validate(self, pipeline: Dict) -> ValidationResponse:
+        """
+        Validates the pipeline JSON payload
+        :param pipeline: the pipeline definition to be validated
+        :return: ValidationResponse containing any and all issues discovered during the validation
+        """
+
+        response = ValidationResponse()
+        root_dir = self.root_dir
+
+        try:
+            pipeline_json = json.loads(json.dumps(pipeline))
+        except (ValueError, json.JSONDecodeError):
+            response.add_message(severity=ValidationSeverity.Error,
+                                 message_type="invalidJSON",
+                                 message="Invalid JSON detected, unable to continue.")
+            return response
+
+        pipeline_execution = pipeline_json['pipelines'][0]['app_data'].get('runtime')  # local, kfp, airflow
+        pipeline_runtime = self._get_runtime_schema(pipeline, response)
+
+        self._validate_pipeline_structure(pipeline=pipeline, response=response)
+        await self._validate_compatibility(pipeline=pipeline, response=response,
+                                           pipeline_runtime=pipeline_runtime, pipeline_execution=pipeline_execution)
+        await self._validate_node_properties(root_dir=root_dir, pipeline=pipeline,
```

Comment:
`self._validate_node_properties()` should get `root_dir` from `self.root_dir`.

Reply:
fixed